specifications_dbus.md

```diff
@@ -0,0 +1,92 @@
+
+## Connector
+
+A *connector* is an application, or software library used in an application, which wants to receive push messages from a UnifiedPush distributor.
+
+When a connector first wants to register, it should fetch all service names currently registered on the session bus, and filter it to only those which begin with `org.unifiedpush.Distributor.`. It should then choose one of these (see [Appendix: Implementation practices](#appendix-implementation-practices)) to use as its distributor. It must then generate a random string to use as a token, and call the [org.unifiedpush.Distributor1.Register](#orgunifiedpushdistributor1register-string--string-string) method with this token as the argument. If the return value indicates "NEW_ENDPOINT", the connector must then save the token and the endpoint received in its persistent storage. Registration is now complete.
+
+On subsequent startups, the connector should call [org.unifiedpush.Distributor1.Register](#orgunifiedpushdistributor1register-string--string-string) with the *same* token, and store the new endpoint returned, if it has changed. The connector should also implement the [org.unifiedpush.Connector1.NewEndpoint](#orgunifiedpushconnector1newendpoint-string-string--nothing) method to receive notification about a changed endpoint from the distributor at any time.
+
+The application SHOULD provide an activatable D-Bus service, so that it can be activated by a new push message when it is not running. See [Appendix: Implementation practices](#appendix-implementation-practices).
+
+## Connector API
+
+Connectors MUST provide a service with their application ID, that is, the same name as the name that they sent their `org.unifiedpush.Distributor1.Register` call from.
+
+Connectors MUST implement the `org.unifiedpush.Connector1` interface at the object path `/org/unifiedpush/Connector`. The caller of the methods in this interface MUST NOT wait for a response from them!
+
+### org.unifiedpush.Connector1.Message (String, String, String) → nothing
+
+The distributor calls this method to send a new push message to the connector.
+
+The first argument is the token of the connection, the second argument is the raw POST data as received by the distributor, and the third argument is optionally an ID identifying the message (otherwise it is an empty string). This method does not return anything.
+
+### org.unifiedpush.Connector1.NewEndpoint (String, String) → nothing
+
+The distributor calls this method to inform the connector of the endpoint URL, both after registering for the first time, and if the endpoint changes afterwards. In addition to implementing this method, the connector SHOULD also call [org.unifiedpush.Distributor1.Register](#orgunifiedpushdistributor1register-string--string-string) on every startup to request the newest endpoint URL, in case it missed this call otherwise.
```

Comment:
The distributor MUST call this method to inform the connector of the endpoint URL, both after registering for the first time, and if the endpoint changes afterwards. 

This is best practice : In addition to implementing this method, the <del>connector</del> end user application SHOULD also call [org.unifiedpush.Distributor1.Register](#orgunifiedpushdistributor1register-string--string-string) on every startup to request the newest endpoint URL, in case it missed this call otherwise.